aten/src/ATen/native/cpu/ActivationKernel.cpp

```diff
@@ -0,0 +1,484 @@
+    }
+  }
+  tmpsum.store(tmp_sum_arr);
+  for (int64_t i = 0; i < Vec::size; i++) {
+    tmp_sum_scalar += tmp_sum_arr[i];
+  }
+  return tmp_sum_scalar;
+}
+
+template <typename scalar_t>
+inline void _vec_log_softmax_lastdim(
+    scalar_t* input_data_base,
+    scalar_t* output_data_base,
+    int64_t outer_size,
+    int64_t dim_size) {
+  using Vec = vec256::Vec256<scalar_t>;
+  static constexpr int64_t CHUNK_SIZE = (128 / sizeof(scalar_t)) * Vec::size;
+  int64_t grain_size = internal::TBB_GRAIN_SIZE / (dim_size * CHUNK_SIZE);
+  if (grain_size < CHUNK_SIZE)
+    grain_size = CHUNK_SIZE;
+  //std::cout << "CHUNK_SIZE: " << CHUNK_SIZE<< std::endl;
+  //std::cout << "Vec::size: " << Vec::size << std::endl;
+  //std::cout << "outer_size: " << outer_size<< std::endl;
+
+  tbb::parallel_for(
+      tbb::blocked_range<int64_t>(0, outer_size, grain_size),
+      [&](const tbb::blocked_range<int64_t>& r) {
+        for (int64_t ii = r.begin(); ii < r.end(); ii += CHUNK_SIZE) {
+          scalar_t tmp_sum_scalar[CHUNK_SIZE];
+          scalar_t max_input_arr[CHUNK_SIZE];
+          int64_t inner_loop_end = CHUNK_SIZE;
+          if (ii + CHUNK_SIZE > r.end())
+            inner_loop_end = r.end() - ii;
+          //std::cout<< "inner_loop_end: " << inner_loop_end << std::endl;
+          for (int64_t j = 0; j < inner_loop_end; j++) {
+            int64_t i = ii + j;
+            scalar_t* input_data = input_data_base + i * dim_size;
+            max_input_arr[j] = _vec_max(dim_size, input_data);
+            //std::cout << "max_input_arr[" << j << "]: " << max_input_arr[j] << std::endl;
+          }
+          for (int64_t j = 0; j < inner_loop_end; j++) {
+            int64_t i = ii + j;
+            tmp_sum_scalar[j] = 0;
+            scalar_t* input_data = input_data_base + i * dim_size;
+            Vec tmpsum(0);
+            Vec max_input_vec(max_input_arr[j]);
+            for (int64_t d = 0; d < dim_size; d += Vec::size) {
+              Vec value(0);
+              int64_t leftover = _leftover<Vec::size>(d, dim_size);
+              //std::cout << "d: " << d << " - leftover: " << leftover << std::endl;
+              value.load(input_data + d, leftover);
+              value = value - max_input_vec;
+              value = value.exp();
+              // Need to do a partial add
+              if (leftover < Vec::size) {
+                scalar_t value_arr[Vec::size];
+                value.store(value_arr);
+                for (int64_t i = 0; i < leftover; i++)
+                  tmp_sum_scalar[j] += value_arr[i];
+              } else {
+                tmpsum = tmpsum + value;
+              }
+            }
+            scalar_t tmp_sum_arr[Vec::size];
+            tmpsum.store(tmp_sum_arr);
+            for (int64_t i = 0; i < Vec::size; i++) 
+              tmp_sum_scalar[j] += tmp_sum_arr[i];
+            //std::cout << "tmp_sum_scalar[" << j << "]: " << tmp_sum_scalar[j] << std::endl;
+          }
```

Comment:
And we have our `vec_reduce` appear once again!

Reply:
Yes